pkg/roll/run_update.go

```diff
@@ -87,24 +88,15 @@ func RetryOrQuit(f func() error, quit <-chan struct{}, logger logging.Logger, er
 func (u update) Run(quit <-chan struct{}) (ret bool) {
 	u.logger.NoFields().Debugln("Locking")
 	// TODO: implement API for blocking locks and use that instead of retrying
-	if !RetryOrQuit(u.lockRCs, quit, u.logger, "Could not lock update") {
+	if !RetryOrQuit(
+		func() error { return u.lockRCs(quit) },
+		quit,
+		u.logger,
+		"Could not lock update",
+	) {
 		return
 	}
-
-	// release both locks on termination
-	defer func() {
-		// failure to unlock probably means:
-		// - session was already invalidated, lock was already released when
-		//   that happened
-		// - network error, session will probably time out soon
-		// either way no point retrying
-		if err := u.lock.Unlock(u.lockPath(u.OldRC)); err != nil {
-			u.logger.WithError(err).Warnln("Could not unlock old RC for updates")
-		}
-		if err := u.lock.Unlock(u.lockPath(u.NewRC)); err != nil {
-			u.logger.WithError(err).Warnln("Could not unlock new RC for updates")
-		}
-	}()
+	defer u.unlockRCs(quit)
```

Comment:
I was concerned about it running forever but you've addressed that in another thread :+1: